x-pack/filebeat/tests/integration/managerV2_test.go

```diff
@@ -500,6 +528,157 @@ func TestRecoverFromInvalidOutputConfiguration(t *testing.T) {
+		},
+		{
+			Id:             "input-unit-1",
+			Type:           proto.UnitType_INPUT,
+			ConfigStateIdx: 1,
+			State:          proto.State_HEALTHY,
+			LogLevel:       logLevel,
+			Config: &proto.UnitExpectedConfig{
+				Id:   "filestream-monitoring-agent",
+				Type: "filestream",
+				Name: "filestream-monitoring-agent",
+				Streams: []*proto.Stream{
+					{
+						Id: "log-input-1",
+						Source: integration.RequireNewStruct(t, map[string]interface{}{
+							"enabled": true,
+							"type":    "log",
+							"paths":   []interface{}{logFilePath},
+						}),
+					},
+				},
+			},
+		},
+	}
+	agentInfo := proto.AgentInfo{
+		Id:       "agent-id",
+		Version:  wantVersion,
+		Snapshot: true,
+	}
+
+	server := &mock.StubServerV2{
+		CheckinV2Impl: func(observed *proto.CheckinObserved) *proto.CheckinExpected {
+			return &proto.CheckinExpected{
+				AgentInfo: &agentInfo,
+				Units:     units,
+			}
+		},
+		ActionImpl: func(response *proto.ActionResponse) error { return nil },
+	}
+
+	rootKey, rootCACert, rootCertPem, err := certutil.NewRootCA()
+	require.NoError(t, err, "could not generate root CA")
+
+	rootCertPool := x509.NewCertPool()
+	ok := rootCertPool.AppendCertsFromPEM(rootCertPem)
+	require.Truef(t, ok, "could not append certs from PEM to cert pool")
+
+	beatPrivKeyPem, beatCertPem, beatTLSCert, err :=
+		certutil.GenerateChildCert("localhost", rootKey, rootCACert)
+	require.NoError(t, err, "could not generate child TLS certificate")
+
+	getCert := func(info *tls.ClientHelloInfo) (*tls.Certificate, error) {
+		// it's one of the child certificates. As there is only one, return it
+		return beatTLSCert, nil
+	}
+
+	creds := credentials.NewTLS(&tls.Config{
+		ClientAuth:     tls.RequireAndVerifyClientCert,
+		ClientCAs:      rootCertPool,
+		GetCertificate: getCert,
+		MinVersion:     tls.VersionTLS12,
+	})
+	err = server.Start(grpc.Creds(creds))
+	require.NoError(t, err, "failed starting GRPC server")
+	t.Cleanup(server.Stop)
+
+	filebeat.Start("-E", "management.enabled=true")
+
+	startUpInfo := &proto.StartUpInfo{
+		Addr:       fmt.Sprintf("localhost:%d", server.Port),
+		ServerName: "localhost",
+		Token:      "token",
+		CaCert:     rootCertPem,
+		PeerCert:   beatCertPem,
+		PeerKey:    beatPrivKeyPem,
+		Services:   []proto.ConnInfoServices{proto.ConnInfoServices_CheckinV2},
+		AgentInfo:  &agentInfo,
+	}
+	writeStartUpInfo(t, filebeat.Stdin(), startUpInfo)
+	// for some reason the pipe needs to be closed for filebeat to read it.
+	require.NoError(t, filebeat.Stdin().Close(), "failed closing stdin pipe")
+
+	msg := strings.Builder{}
+	require.Eventuallyf(t, func() bool {
+		msg.Reset()
+
+		_, err = os.Stat(eventsDir)
+		if err != nil {
+			fmt.Fprintf(&msg, "could not verify output directory exists: %v", err)
+			return false
+		}
+
+		entries, err := os.ReadDir(eventsDir)
+		if err != nil {
+			fmt.Fprintf(&msg, "failed checking output directory for files: %v", err)
+			return false
+		}
+
+		if len(entries) == 0 {
```

Comment:
It looks like there is a race condition here where the file could exist but be empty, with the events being written in later.

Can you check that the length of the file is not zero as well?

Reply:
done [here](https://github.com/elastic/beats/pull/37553/files#diff-df9d93045b24b2f8f834d399fb096632e419254901cb9f0c504636c7662c369dR685)

Reply:
Technically this is still a race, because you don't need at least 1 byte, you need at least one line.

Reply:
You may be better off calling getEventsFromFileOutput in a require.Eventually loop.

Reply:
You could keep the `i.Size() == 0` where it is, and then try `getEventsFromFileOutput` a few times. What you want to do is avoid writing 50% of a valid event which will fail with a JSON parsing error.